core/FileHash.cc

```diff
@@ -13,6 +13,44 @@ uint32_t incZero(uint32_t a) {
     return a == 0 ? 1 : a;
 };
 
+// We ignore unique names because those are names created by Sorbet not written by the user. If a
+// symbol with a unique name changes (like an overload, or a T::Enum name, or a Singleton class), we
+// want to force the fast path to retypecheck any file that mentioned the original name in any way
+//
+// For example, `A#foo (overload.1)` changes, that means we should typecheck all files with `foo`,
+// because the that `foo` might end up dispatching to the unique name, despite not mentioning the
+// unique name directly.
+//
+// Similarly for constant names--changes to the the fake `MyEnum::X$1` symbols we create to model
+// `T::Enum` subclasses could affect a file that simply mentions a constant with the name `X` (but
+// not a file that mentions a method called `X()`, which is why we still include the name kind in
+// the hash for non-UNIQUE names).
+uint32_t hashNameRefWithoutUniques(const GlobalState &gs, NameRef nm) {
+    uint32_t result;
+    auto kind = nm.kind();
+
+    switch (kind) {
+        case NameKind::UTF8: {
+            result = _hash(nm.dataUtf8(gs)->utf8);
+            break;
+        }
+        case NameKind::CONSTANT: {
+            result = hashNameRefWithoutUniques(gs, nm.dataCnst(gs)->original);
+            break;
+        }
+        case NameKind::UNIQUE: {
+            result = hashNameRefWithoutUniques(gs, nm.dataUnique(gs)->original);
+            break;
+        }
+    }
+
+    if (kind != NameKind::UNIQUE) {
+        auto hashedKind = static_cast<underlying_type<NameKind>::type>(kind);
+        result = mix(result, hashedKind);
+    }
+    return result;
+}
```

Comment:
This code follows the structure of `hashFullNameRef` and consistency is a good thing, so the below is only a suggestion.

Probably slightly better to structure this such that the `hashNameRefWithoutUniques` calls appear in tail position.  So either duplicating the `kind != NameKind::UNIQUE` conditional into the appropriate cases, or moving the conditional before the `switch` (and making sure `result` is initialized in the `UNIQUE` case).

I think the code duplication is actually slightly better; the compiler should be able to constant-fold away the `mix` on the `NameKind` (since `NameKind` will be a constant in each case of the switch), so the performance should be better.